### ISSUE TITLE
Accessing the neuron statistics within it's custom surrogate function

### ISSUE THREAD

**Issue by gwgknudayanga** [fangwei123456/spikingjelly#281]
hi,
Is it possible to access the parameters relevant to the neuron within it surrogate function?
For example, in this following code within the surrogate function , i want to know what is the current voltage and the current time constant of the ParametricLIF neuron 'mylif'. How those paramers can be queried within its surrogate_function? Thanks !

mylif = ParametricLIFNode(surrogate_function=STE())

class ste(torch.autograd.Function):
@staticmethod
def forward(ctx, x, alpha):
if x.requires_grad:
ctx.save_for_backward(x)
ctx.alpha = alpha
return surrogate.heaviside(x)

@staticmethod
def backward(ctx, grad_output):
    x = ctx.saved_tensors[0]
    mask = (x.abs() <= ctx.alpha).to(x)
    return grad_output * mask, x.abs().mean()

class STE(nn.Module):
def init(self, alpha: float = 1.):
super(STE, self).init()
self.alpha = nn.Parameter(torch.as_tensor(alpha))

def forward(self, x):
    return ste.apply(x, self.alpha)

**Comment by fangwei123456** [comment]
> How those paramers can be queried within its surrogate_function?

This should be implemented by the neuron.

Here is an example:

```python
class ParametricLIFNode(neuron.ParametricLIFNode):
    def neuronal_charge(self, x: torch.Tensor):
        print('v before neuronal_charge', self.v)
        super().neuronal_charge(x)
        print('v after neuronal_charge', self.v)
```

Note that you shoule use `torch` backend. These changed codes will not effert `cupy` backend.

**Comment by gwgknudayanga** [comment]
Thanks Fang. So now i can pass the value of current tau to the surrogate function within the neuronal_charge function of the neuron as follows. Isn't it? And since i am using the single step mode, the backend is always torch. 
import torch
import torch.nn as nn
from spikingjelly.activation_based import neuron, surrogate

class ParametricLIFNode(neuron.ParametricLIFNode):
    def neuronal_charge(self, x: torch.Tensor):
        print('v before neuronal_charge', self.v)
        super().neuronal_charge(x)
        print('v after neuronal_charge', self.v)
        self.surrogate_function.getcurrentTauOfNeuron(self.w)
        
class ste(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x, tau):
        if x.requires_grad:
            ctx.save_for_backward(x)
            ctx.tau = tau
        return surrogate.heaviside(x)

    @staticmethod
    def backward(ctx, grad_output):
        x = ctx.saved_tensors[0]
        mask = (x.abs() <= ctx.tau).to(x)
        return grad_output * mask, None


class STE(nn.Module):
    def __init__(self, tau: float = 1.):
        super(STE, self).__init__()
        self.tau = 0

    def getcurrentTauOfNeuron(self,tau):
        self.tau = tau

    def forward(self, x):
        return ste.apply(x, self.tau)

net = ParametricLIFNode(surrogate_function=STE())

x = torch.rand([4, 8], requires_grad=True)

net(x).sum().backward()
print(x.grad)

**Comment by fangwei123456** [comment]
? So now i can pass the value of current tau to the surrogate function within the neuronal_charge function of the neuron as follows. Isn't it?

Yes. You can also print `v` (or other attributes like `tau`) before and after fire by mimicing codes in 
https://github.com/fangwei123456/spikingjelly/issues/281#issuecomment-1278973925

> And since i am using the single step mode, the backend is always torch.

Yes.

**Comment by gwgknudayanga** [comment]
> ? So now i can pass the value of current tau to the surrogate function within the neuronal_charge function of the neuron as follows. Isn't it?
> 
> Yes. You can also print `v` (or other attributes like `tau`) before and after fire by mimicing codes in [#281 (comment)](https://github.com/fangwei123456/spikingjelly/issues/281#issuecomment-1278973925)
>  Thanks . Actually i used parameter 'w' because when i use tau, it gives error saying that "'ParametricLIFNode' object has no attribute 'tau'".
> > And since i am using the single step mode, the backend is always torch.
> 
> Yes.

**Comment by fangwei123456** [comment]
> Actually i used parameter 'w' because when i use tau, it gives error saying that "'ParametricLIFNode' object has no attribute 'tau'".

In PLIF, `tau = self.w.sigmoid()`.

https://github.com/fangwei123456/spikingjelly/blob/f8014142a0fbb7d05610fc76412784295ffde71b/spikingjelly/activation_based/neuron.py#L1063

**Comment by gwgknudayanga** [comment]
> > Actually i used parameter 'w' because when i use tau, it gives error saying that "'ParametricLIFNode' object has no attribute 'tau'".
> 
> In PLIF, `tau = 1 / self.w.sigmoid()`.
> 
> https://github.com/fangwei123456/spikingjelly/blob/f8014142a0fbb7d05610fc76412784295ffde71b/spikingjelly/activation_based/neuron.py#L1063

Thanks Fang !!
